File: app/src/main/java/top/niunaijun/blackboxa/util/CrashHandler.java
package top.niunaijun.blackbox.util;

import android.content.Context;

import java.io.File;
import java.io.FileWriter;
import java.io.PrintWriter;

public class CrashHandler implements Thread.UncaughtExceptionHandler {

    private static CrashHandler instance = new CrashHandler();
    private Thread.UncaughtExceptionHandler defaultHandler;
    private Context context;

    public static CrashHandler getInstance() {
        return instance;
    }

    public void init(Context ctx) {
        context = ctx;
        defaultHandler = Thread.getDefaultUncaughtExceptionHandler();
        Thread.setDefaultUncaughtExceptionHandler(this);
    }

    @Override
    public void uncaughtException(Thread thread, Throwable throwable) {

        try {
            File dir = new File(context.getExternalFilesDir(null), "crash");
            if (!dir.exists()) dir.mkdirs();

            File file = new File(dir, "crash.txt");

            PrintWriter pw = new PrintWriter(new FileWriter(file, true));
            pw.println("===== Crash =====");
            throwable.printStackTrace(pw);
            pw.close();

        } catch (Exception ignored) {}

        if (defaultHandler != null) {
            defaultHandler.uncaughtException(thread, throwable);
        }
    }
}

File: app/src/main/java/top/niunaijun/blackboxa/util/CrashHandler.kt
package top.niunaijun.blackboxa.util

import android.content.Context
import android.util.Log
import java.io.File
import java.io.FileWriter
import java.lang.Thread.UncaughtExceptionHandler
import java.text.SimpleDateFormat
import java.util.*

class CrashHandler private constructor() : UncaughtExceptionHandler {

    private var defaultHandler: UncaughtExceptionHandler? = null
    private lateinit var logDir: File

    companion object {
        private val instance = CrashHandler()
        private const val TAG = "CrashHandler"

        fun get(): CrashHandler = instance
    }

    fun init(context: Context) {

        // Safer path for Android 11+
        logDir = File("/storage/emulated/0/Android/data/top.niunaijun.blackboxa/crash_logs")

        if (!logDir.exists()) {
            logDir.mkdirs()
        }

        defaultHandler = Thread.getDefaultUncaughtExceptionHandler()
        Thread.setDefaultUncaughtExceptionHandler(this)

        Log.d(TAG, "CrashHandler initialized. Log dir: ${logDir.absolutePath}")
    }

    override fun uncaughtException(thread: Thread, throwable: Throwable) {
        val logFile = saveCrashLog(thread, throwable)
        Log.e(TAG, "App crashed! Log saved to: $logFile", throwable)

        defaultHandler?.uncaughtException(thread, throwable)
    }

    private fun saveCrashLog(thread: Thread, throwable: Throwable): File {

        val timestamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault()).format(Date())
        val fileName = "crash_$timestamp.log"

        val logFile = File(logDir, fileName)

        val crashReport = buildString {
            appendLine("========= CRASH LOG =========")
            appendLine("Time: ${SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault()).format(Date())}")
            appendLine("Thread: ${thread.name} (ID: ${thread.id})")
            appendLine("Exception: ${throwable.javaClass.name}")
            appendLine("Message: ${throwable.message}")
            appendLine()
            appendLine("--------- STACK TRACE ---------")
            appendLine(throwable.stackTraceToString())
            appendLine("========= END =========")
        }

        try {
            FileWriter(logFile).use { it.write(crashReport) }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save crash log", e)
        }

        return logFile
    }

    fun getLogFiles(): List<File> {
        return logDir.listFiles()?.filter { it.extension == "log" }?.toList() ?: emptyList()
    }

    fun getLatestLog(): File? {
        return getLogFiles().maxByOrNull { it.lastModified() }
    }
}
